### DISCUSSION TITLE
Tree-shaking support for module federation

### DISCUSSION THREAD

**Discussion by XiaofengXie16** [webpack/webpack#12532] [Q&A]
Is there going to be a plan for adding tree shaking support for module federation?

Use case:
Several of my apps are sharing 10-15 ui components from a large UI library (tree-shakable) . However, when I try to use module federation to share this library, I would have to share the entire UI library instead of the components that the apps use which means I have to download unused code from the library.

**Comment by alexander-akait** [comment]
Because we don't know which components you will use, how you imagine it?

**Reply by literalpie** [reply]
It would be nice if there was a way to define all expected possible imports, and maybe webpack could use that to tree-shake? Maybe this could just be a js file with a list of import statements.

It would be even cooler if, when code references something not in the "expected imports", it could fall back to the un-treeshaken version of the dependency.

It would absolutely awesome if there was a way in dev mode to "trace" all imports that happen as you run the app and automatically generate the "all needed imports definition" mentioned in step 1.

(I'm just spit-balling here - sorry if none of this is feasible)

**Reply by literalpie** [reply]
I played around with this some and found some things that might be useful. I made [a commit in this fork of the examples repo](https://github.com/literalpie/module-federation-examples/commit/0d15f853945caf340644cee73606bbd0069c25b2) with my notes and example code.

The gist is that, in my experimenting, you can "trick" webpack into treeshaking in certain ways by having a shared dependency that re-exports the specific things that you want included.

I'm always impressed whenever I say "it would be cool if webpack could..." and it actually does it!

**Reply by rodoabad** [reply]
You can still tree shake your library upon consumption for your remotes. To which I guess, it's going to start to depend whether it's better to add that package to `shared` versus letting Webpack tree shake that package upon consumption of your remotes.

@alexander-akait would you agree on that approach instead?

**Reply by alexander-akait** [reply]
@rodoabad Yeah, I think it already works as you written
